### PR TITLE
Refactor logging to use unique id per request and take logging level from user

### PR DESCRIPTION
**Pull request by tmccrmck**

### CODE REVIEW

File: onnxruntime/hosting/environment.cc [hunk 1]
```diff
@@ -10,10 +10,10 @@
 namespace onnxruntime {
 namespace hosting {
 
-HostingEnvironment::HostingEnvironment() : logger_id_("HostingLog"),
+HostingEnvironment::HostingEnvironment(onnxruntime::logging::Severity severity) : logger_id_("HostingLog"),
```
Comment (NonStatic2014) [inline onnxruntime/hosting/environment.cc]: Can you change `HostingLog` to `Hosting` or `HostingApp`?

File: onnxruntime/hosting/environment.cc [hunk 2]
```diff
@@ -23,13 +23,17 @@ HostingEnvironment::HostingEnvironment() : logger_id_("HostingLog"),
   session_ = std::make_shared<onnxruntime::InferenceSession>(options_, &default_logging_manager_);
 }
 
-const onnxruntime::logging::Logger& HostingEnvironment::GetLogger() {
+const onnxruntime::logging::Logger& HostingEnvironment::GetDefaultLogger() {
```
Comment (NonStatic2014) [inline onnxruntime/hosting/environment.cc]: Make it `GetAppLogger()`? Then when someone write code s\he will not confuse what is the default logger.

File: onnxruntime/hosting/environment.cc [hunk 3]
```diff
@@ -23,13 +23,17 @@ HostingEnvironment::HostingEnvironment() : logger_id_("HostingLog"),
...
   return this->default_logging_manager_.DefaultLogger();
 }
 
+std::shared_ptr<onnxruntime::logging::Logger> HostingEnvironment::GetLogger(const std::string& id) {
+  return this->default_logging_manager_.CreateLogger(id);
```
Comment (NonStatic2014) [inline onnxruntime/hosting/environment.cc:35]: If the `id` is empty string, we should log an warning.

File: onnxruntime/hosting/http/context.h [hunk 4]
```diff
@@ -1,10 +1,20 @@
...
+#define ONNXRUNTIME_HOSTING_HTTP_CONTEXT_H
+
+// boost random is using a deprecated header in 1.69
+// See: https://github.com/boostorg/random/issues/49
+#define BOOST_PENDING_INTEGER_LOG2_HPP
+#include <boost/integer/integer_log2.hpp>
+
+#include <string>
 
 #include <boost/beast/http.hpp>
```
Comment (NonStatic2014) [inline onnxruntime/hosting/http/context.h:14]: Another option would be libuuiddev, which need the `sudo apt-get install uuid-dev`

Here is an quick example: https://gist.github.com/yoggy/4483031

File: onnxruntime/hosting/http/predict_request_handler.cc [hunk 5]
```diff
@@ -15,11 +15,11 @@ void Predict(const std::string& name,
...
   PredictRequest predictRequest{};
-  auto logger = env->GetLogger();
+  auto logger = env->GetLogger(context.uuid);
 
-  LOGS(logger, VERBOSE) << "Name: " << name;
-  LOGS(logger, VERBOSE) << "Version: " << version;
-  LOGS(logger, VERBOSE) << "Action: " << action;
+  LOGS(*logger, VERBOSE) << "Name: " << name;
+  LOGS(*logger, VERBOSE) << "Version: " << version;
+  LOGS(*logger, VERBOSE) << "Action: " << action;
```
Comment (NonStatic2014) [inline onnxruntime/hosting/http/predict_request_handler.cc]: NIT: we could put them in one line.

File: onnxruntime/hosting/http/context.h (hunk 4 above)
Comment (NonStatic2014) [reply onnxruntime/hosting/http/context.h:14]: On Windows, using `CoCreateGuid()`

File: onnxruntime/hosting/server_configuration.h [hunk 6]
```diff
@@ -9,21 +9,31 @@
...
 // Wrapper around Boost program_options and should provide all the functionality for options parsing
 // Provides sane default values
 class ServerConfiguration {
  public:
+  const std::string full_desc = "ONNX Hosting: host an ONNX model for inferencing with ONNXRuntime";
+  std::string model_path;
+  std::string address = "0.0.0.0";
+  int http_port = 8001;
+  int num_http_threads = std::thread::hardware_concurrency();
+  onnxruntime::logging::Severity logging_level{};
```
Comment (NonStatic2014) [inline onnxruntime/hosting/server_configuration.h:32]: We should set a default one, e.g. warning.

File: onnxruntime/hosting/server_configuration.h [hunk 7]
```diff
@@ -51,19 +61,22 @@ class ServerConfiguration {
...
+  po::options_description desc{"Allowed options"};
+  po::variables_map vm{};
+  std::string logging_level_str = "verbose";
+
   // Print help and return if there is a bad value
   Result ValidateOptions() {
-    if (num_http_threads <= 0) {
+    if (vm.count("logging_level") &&
+        (!(logging_level_str == "verbose" || logging_level_str == "info") || logging_level_str == "warning" || logging_level_str == "error" || logging_level_str == "fatal")) {
+      PrintHelp(std::cerr, "logging_level must be one of verbose, info, warning, error, or fatal");
```
Comment (NonStatic2014) [inline onnxruntime/hosting/server_configuration.h:77]: We should mention we are case sensitive.

File: onnxruntime/hosting/server_configuration.h [hunk 8]
```diff
@@ -82,6 +95,15 @@ class ServerConfiguration {
     return vm.count("help") || vm.count("h");
   }
 
+  onnxruntime::logging::Severity GetSeverity(const std::string& level) {
+    if (level == "verbose") return onnxruntime::logging::Severity::kVERBOSE;
```
Comment (NonStatic2014) [inline onnxruntime/hosting/server_configuration.h:99]: Can we use an `unordered_map` here?

File: tools/ci_build/build.py [hunk 9]
```diff
@@ -227,7 +227,7 @@ def install_hosting_deps(source_dir):
         boostrap_path = os.path.join(vcpkg_folder_path, 'bootstrap-vcpkg' + file_ending)
         run_subprocess([boostrap_path])
 
-    run_subprocess([vcpkg_executable, 'install', 'boost-beast', 'boost-program-options', 'rapidjson'])
+    run_subprocess([vcpkg_executable, 'install', 'boost-beast', 'boost-program-options', 'boost-uuid'])
```
Comment (NonStatic2014) [inline tools/ci_build/build.py:230]: If we could use libuuiddev, we may get rid of this new dependency.

File: onnxruntime/hosting/server_configuration.h (hunk 8 above)
Comment (tmccrmck) [reply onnxruntime/hosting/server_configuration.h:99]: As we said offline there's no reason to do this

File: onnxruntime/hosting/server_configuration.h (hunk 6 above)
Comment (tmccrmck) [reply onnxruntime/hosting/server_configuration.h:32]: We do but the string field is private so it's not here and the enum defaults to 0.

File: tools/ci_build/build.py (hunk 9 above)
Comment (tmccrmck) [reply tools/ci_build/build.py:230]: As discussed we may change this with feedback from onnxruntime team

File: onnxruntime/hosting/http/context.h (hunk 4 above)
Comment (tmccrmck) [reply onnxruntime/hosting/http/context.h:14]: Leaving boost-uuid for now

File: onnxruntime/hosting/environment.cc [hunk 10]
```diff
@@ -28,6 +28,10 @@ const onnxruntime::logging::Logger& HostingEnvironment::GetAppLogger() {
 }
 
 std::shared_ptr<onnxruntime::logging::Logger> HostingEnvironment::GetLogger(const std::string& id) {
+  if (id.empty()) {
+    LOGS(GetAppLogger(), VERBOSE) << "Request id is null or empty string";
```
Comment (NonStatic2014) [inline onnxruntime/hosting/environment.cc]: Make it `WARNING`?

**Review by NonStatic2014 — APPROVED** [review]

**Pull request merged**